jsonrpc: don't crash on multiple commands at once.

rustyrussell opened the pull request:
Once we read a command, we are supposed to io_wait until it finishes.
However, we are actually woken in two places: when it's complete
(which is correct), and when it's written out (which is wrong).

~~~This was because the reader and writer were waiting on the same location: make the writer wait/wake on the output queue, instead.~~~
Indeed, when a command is finished, it's time to wake both the output and the input: 
by making the low-level `json_done` routine free the old cmd as well as queue the output,
we reduce code duplication and also make it clear what it's doing.

Fixes: #934
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ZmnSCPxj commented:
Code modification looks OK to me (I must confess I have not dipped into the `io_wait*` system yet and am commenting here from a position of ignorance).

Could we also add this sequence to the test?

https://github.com/ElementsProject/lightning/issues/934#issuecomment-364274318

    { "method": "invoice", "params": [100, "foo", "foo"], "jsonrpc": "2.0", "id": 0 }
    { "method": "waitinvoice", "params": ["foo"], "jsonrpc" : "2.0", "id": 1 }
    { "method": "delinvoice", "params": ["foo", "unpaid"], "jsonrpc" : "2.0", "id": 2 }

The above should get 3 objects from the readback to ensure that the `waitinvoice` is correctly handled.

While client applications are unlikely to use `waitinvoice` in the above manner, the same issue underlies use of `waitanyinvoice` and an application using only a single JSON-RPC connection, to make all invoice-related calls including `waitanyinvoice` and `invoice` and so on:

1.  Application initializes a single JSON-RPC connection to c-lightning.
2.  Application launches a waiter thread/greenthread/process/Promise which just does `waitanyinvoice` on that JSON-RPC connection and handles each invoice as it comes in.
3.  Application issues invoices (`invoice` command) to payers on the main thread, using the same JSON-RPC connection.
4.  JSON-RPC commands and replies do not get mixed up, because the application uses a library that allocates `id` uniquely and sends the correct reply to the correct place, so the application uses a single JSON-RPC connection for all requests.

cdecker commented:
I tested the case that @ZmnSCPxj proposed, and indeed it manages to crash the daemon by crossing two requests (pending `waitinvoice` is overtaken `delinvoice`) and then we get the JSON results crossed. Adding the test case to this PR.

rustyrussell commented:
Nice!  Thanks, I'll get on testing this one today...

rustyrussell commented:
Got it.  My fix was not correct, I've rebased and fixed it properly.   I also added a fixup which makes @cdecker's test not hang, now it doesn't crash!

ZmnSCPxj commented:
NACK bc1ba13f0a4892a063403e550f8e58edb6ce9cfa

The point of the test is that currently `waitinvoice` prevents future commands from executing.  But, as I mentioned in https://github.com/ElementsProject/lightning/pull/994#issuecomment-365513479 , it is reasonable for an application to open a single JSON-RPC connection, issue a `waitanyinvoice` which it waits for to handle incoming payments, and expect to reuse the same connection for future calls to `invoice` et al to manage invoices.  By putting the timeout here to 1 second we are simply sidestepping the point of the test, which is that the later `delinvoice` should be what resumes the `waitinvoice` in this case, not the expiration of the invoice.

rustyrussell commented:
I see your point, but that's a completely separate change, and orthogonal to this one.

I've opened a new issue, as this one correctly fixes the crash and contains a valid test, I think we should apply this now.

cdecker commented:
ACK bc1ba13
